### PR TITLE
Flamer changes

### LINKED ISSUE #5598: Delayed ammo counters
Ammo counters show old ammo values in some instances:

1) When tactical reloading, the ammo counter does not update until it is refreshed by shooting or unwielding and wielding again. For example, if your gun has 20 bullets left and you tactical reload a 50 magazine in, the ammo counter will still display 20 rounds. Shooting will refresh it to the correct value, but this can get confusing if you're keeping mags that aren't full on you (Happened once when mags for a gun were almost all vended out but other mags with the same ammo were around, so I had to keep them so as to manually load them at fob).

2) TL-84 ammo counter will subtract the ammo taken by the flame before the current one and display said old value. For example, say you have used a short burst that takes only 2 off the fuel tank. The next flame, you take 5, but given the ammo counter being delayed, it will only visibly subtract 2 and display the old value differing from the fuel tank. This gets confusing because examining the flamethrower only shows the hydrocannon ammo.

### PR DESCRIPTION
**Pull request by BraveMole**
## About The Pull Request

Flamer is now a bit easier to use (auto lighting, and a new keybind to use the hydro cannon). 

Hydro cannon now has a working ammo HUD, it's also much easier to know if you are using it (cause water is not really good against roonies)

Added the back fuel tank, 500 unit of fuel on your back that you can directly use as ammo for your flamer if you link it
Dropping your gun or the fuel tank breaks the link between them, even with magnetic harness or belt harness.

https://vimeo.com/496251683

Fixed #5598, ammo HUD is now up to date when firing any flamie

Copy of  #5607 that I couldn't access anymore

## Why It's Good For The Game

A new option for pyros, it removes the need of changing fuel tanks every so often at the cost of not having any backpack.

I think this new fuel tank is balanced because even though you don't need to reload, if a xenos hit you with a CC and makes you drop your gun, you have to remake the link wich takes the same time as reloading with a traditional flamer.

Having an hydro cannon keybind just makes thing much easier, and the mechanic of lighting your flamer is just here to kill new comers trying to fire their flame thrower the first time.

## Changelog
:cl:
add: Added back fuel tank, can be attached to flamethrowers to act as a huge fuel tank. Multiple versions for different fuel type
add: 2 standard back fuel tanks added in ammunition vendor
add: Req can buy standard back fuel for 20 points and blue flame version for 150 points (Ammo tab)
add: Ammo hud for hydrocannon
tweak: The flamers will now automaticly lit up when they have access to fuel
tweak: You can toggle the hydro cannon of the tl84 using unique action (space bar as defaut)
del: Remove the pyro backpack (replaced by the back fuel tank)
imageadd: Added images of flamers linked to back fuel tank
imageadd: Added water ammo hud
fix: ammo HUD is now up to date after firing a flamethrower
/:cl:

### CODE REVIEW

**Comment by RenGusta** [comment]
Sounds like a neat adjustment of the Flamer and makes it a little more helpful for new users.

**Comment by Chaznoodles** [comment]
Flamer 'magazines' have a timed reload to stop nonstop flamer pressure, this gets around that. I'm not a particular fan of that. We also don't use different flamer fuel types such as the green/blue because of the absolute havoc they wreaked.

Also, you didn't actually make this available anywhere, as far as I can see.

**Comment by BraveMole** [comment]
> Flamer 'magazines' have a timed reload to stop nonstop flamer pressure, this gets around that and means flamers can go hog wild. I can't say I'm a particular fan of that. We also don't use different flamer fuel types such as the green/blue because of the absolute havoc they wreaked.
> 
> Also, you didn't actually make this available anywhere, as far as I can see.

Since dropping the gun unlink the tank with the flamer, a CC is enough to make the flamer innoperant. Think that's enough of disavantadge to make it balanced.

**Comment by SplinterGP** [comment]
the pyro spec kit aint available anywhere, so that fuel tank on back aint either

**Comment by Knuff-les** [comment]
There's already a keybind for the hydro cannon, and unique action is the keybind to light the pilot light for the flamer in the first place

**Comment by BraveMole** [comment]
> There's already a keybind for the hydro cannon, and unique action is the keybind to light the pilot light for the flamer in the first place

There's no keybind currently and the pilot will lit up automaticly

**Comment by Knuff-les** [comment]
> > There's already a keybind for the hydro cannon, and unique action is the keybind to light the pilot light for the flamer in the first place
> 
> There's no keybind currently and the pilot will lit up automaticly

Actually, you're right, There's no native keybind, but you can make it with a macro. Which is SS13 speak for keybind, but that's splitting hairs

File: code/game/objects/items/storage/backpack.dm [hunk 1]
```diff
@@ -756,37 +756,6 @@
 	..()
 	to_chat(user, "[reagents.total_volume] units of fuel left!")
```
Comment (TiviPlus) [inline code/game/objects/items/storage/backpack.dm:757]:
```suggestion
	. = ..()
	to_chat(user, "[reagents.total_volume] units of fuel left!")
```

File: code/game/objects/items.dm [hunk 2]
```diff
@@ -244,6 +244,9 @@
 	if(flags_item & DELONDROP)
 		qdel(src)
 
+///Called whenever an item is unequipped to a new loc (IE, not when the item ends up in the hands)
+/obj/item/proc/dropped_new_loc(mob/user)
+	return
```
Comment (TiviPlus) [inline code/game/objects/items.dm:249]: dropped_from_inventory would be better, no?

File: code/modules/projectiles/guns/flamer.dm [hunk 3]
```diff
@@ -28,27 +28,51 @@
...
+	if (current_mag) //A flamer spawing with a mag will be lit up
+		light_pilot(null,TRUE)
+
 /obj/item/weapon/gun/flamer/unique_action(mob/user)
-	return toggle_flame(user)
+	if(under)
+		var/obj/item/attachable/hydro_cannon/hydro = under
+		if(istype(hydro))
+			hydro.activate_attachment(user)
+			playsound(user, hydro.activation_sound, 15, 1)
```
Comment (TiviPlus) [inline code/modules/projectiles/guns/flamer.dm:41]: early return, and the if(under) isnt needed

File: code/modules/projectiles/guns/flamer.dm [hunk 4]
```diff
@@ -117,37 +136,94 @@
...
 	current_mag = null
-
 	update_icon()
-
 	return TRUE
 
+/obj/item/weapon/gun/flamer/proc/attach_fueltank(mob/user, obj/item/ammo_magazine/flamer_tank/backtank/fueltank)
+	if (!fueltank || !istype(fueltank))
+		to_chat(user, "<span class='warning'>That's not an attachable fuel tank!</span>")
+		return
```
Comment (TiviPlus) [inline code/modules/projectiles/guns/flamer.dm:179]:
```suggestion
	if (!istype(fueltank))
		to_chat(user, "<span class='warning'>That's not an attachable fuel tank!</span>")
		return
```

File: code/modules/projectiles/guns/flamer.dm [hunk 5]
```diff
@@ -117,37 +136,94 @@
...
+		fueltank.attached_flamer = src
+		replace_ammo(user, fueltank)
+		light_pilot(user,TRUE)
+		playsound(user, reload_sound, 25, 1, 5)
+		update_icon(user)
+		var/obj/screen/ammo/A = user.hud_used.ammo
+		A.update_hud(user)
+	else
+		to_chat(user, "<span class='warning'>Your action was interrupted!</span>")
+		return
```
Comment (TiviPlus) [inline code/modules/projectiles/guns/flamer.dm:202]: early return

File: code/modules/projectiles/guns/flamer.dm [hunk 6]
```diff
@@ -117,37 +136,94 @@
...
+		to_chat(user, "<span class='notic'>You detach the fuel tank</span>")
+	playsound(user, unload_sound, 25, 1)
+	light_pilot(user,FALSE)
+	update_icon(user)
+	var/obj/screen/ammo/A = user.hud_used.ammo
+	A.update_hud(user)
+	
+	
+/obj/item/weapon/gun/flamer/dropped_new_loc(mob/user)
+	..()
```
Comment (TiviPlus) [inline code/modules/projectiles/guns/flamer.dm:218]: . = ..()

File: code/modules/projectiles/guns/flamer.dm [hunk 7]
```diff
@@ -117,37 +136,94 @@
...
+	var/mob/living/carbon/human/humanuser = user
+	if (istype(humanuser))
```
Comment (TiviPlus) [inline code/modules/projectiles/guns/flamer.dm:220]: when would it not be a human?

File: code/modules/projectiles/magazines/flamer.dm [hunk 8]
```diff
@@ -48,6 +48,49 @@
...
+				if(FLT.current_mag == src)
+					FLT.detach_fueltank(user)
+				else if (attached_flamer)
+					to_chat(user, "<span class='warning'>This fuel tank is already attached to something</span>")
+				else
+					FLT.attach_fueltank(user,src)
+			else
+				to_chat(user, "<span class='warning'>You must equip or hold this fuel tank to be able to link it to a flamer</span>")
+	else
+		return ..()
```
Comment (TiviPlus) [inline code/modules/projectiles/magazines/flamer.dm:79]: Early returnnnnn

File: code/modules/projectiles/magazines/flamer.dm [hunk 9]
```diff
@@ -48,6 +48,49 @@
...
+
+/obj/item/ammo_magazine/flamer_tank/backtank/dropped_new_loc(mob/user) //Dropping the tank should unlink it from the flamer
+	..()
+	var/mob/living/carbon/human/humanuser = user
+	if (istype(humanuser))
+		if(attached_flamer)
+			attached_flamer.detach_fueltank(user,FALSE)
+	else
+		return ..()
```
Comment (TiviPlus) [inline code/modules/projectiles/magazines/flamer.dm:88]: Duplicate supercalls, early return

**Review by TiviPlus — CHANGES_REQUESTED** [review]

File: code/game/objects/items.dm (hunk 2 above)
Comment (BraveMole) [reply code/game/objects/items.dm:249]: Up for debates, but is it clear that hands are part of inventory? For me it's confusing

Comment (TiviPlus) [reply code/game/objects/items.dm:249]: What do you mean? 
When EXACTLY is this called

File: code/modules/projectiles/guns/flamer.dm (hunk 7 above)
Comment (BraveMole) [reply code/modules/projectiles/guns/flamer.dm:220]: A monkey?

Comment (TiviPlus) [reply code/modules/projectiles/guns/flamer.dm:220]: Monkeys shouldnt be using flamers anyway
They should also be a human species so just remove it

**Review by TheBimmer231 — APPROVED** [review]

**Comment by Walarks** [comment]
I also think that blue flame shouldn't be added. Different flame colours based on what created the flame can be interesting, but in terms of balance currently fire is balanced around there being only 1 fire type and it is generally sufficiently strong.

Blue flame is presently found in incendiary OB to push up its power to that of the other OB types.

File: code/modules/projectiles/magazines/flamer.dm [hunk 10]
```diff
@@ -48,6 +48,53 @@
...
+					return
+				else if (attached_flamer)
+					to_chat(user, "<span class='warning'>This fuel tank is already attached to something</span>")
+					return
+				else
+					FLT.attach_fueltank(user,src)
+					return
+			else
+				to_chat(user, "<span class='warning'>You must equip or hold this fuel tank to be able to link it to a flamer</span>")
+				return
```
Comment (TiviPlus) [inline code/modules/projectiles/magazines/flamer.dm:82]: early return

File: code/modules/projectiles/magazines/flamer.dm [hunk 11]
```diff
@@ -48,6 +48,53 @@
...
+
+/obj/item/ammo_magazine/flamer_tank/backtank/removed_from_inventory(mob/user) //Dropping the tank should unlink it from the flamer
+	. = ..()
+	var/mob/living/carbon/human/humanuser = user
+	if (istype(humanuser))
+		if(attached_flamer)
+			attached_flamer.detach_fueltank(user,FALSE)
+			return
+		return
```
Comment (TiviPlus) [inline code/modules/projectiles/magazines/flamer.dm:91]: ditto

**Review by TiviPlus — CHANGES_REQUESTED** [review]

**Comment by Chaznoodles** [comment]
Blue/green flame were removed because they throw game balance out the window, due to Surts existing. Shouldn't be put back in.

**Comment by Swagile** [comment]
from the testing so far, both being a xeno and as a RO ordering bluefire, i have not seen it do any significant damage to xenos in comparison to just buy a SADAR or minigun with accompanying ammo which is much cheaper

maybe its not being used by good players or something but i haven't seen it be used for good effect, so i wouldn't ever buy it again as a RO 🤷

**Comment by SkeletalElite** [comment]
If you drop your flamer and have a mag harness it unlinks from the fuel

**Comment by BraveMole** [comment]
> If you drop your flamer and have a mag harness it unlinks from the fuel on back

Yeah it's intended, to nerf it a bit. That prevents mindless flamer push that never stop

**Comment by Knuff-les** [comment]
> > If you drop your flamer and have a mag harness it unlinks from the fuel on back
> 
> Yeah it's intended, to nerf it a bit. That prevents mindless flamer push that never stop

I’ve played with this MR, and it’s actually really cool, but that is the reason I personally won’t continue to use this in it’s current state

Usually, fuel loadout is 3 tanks + 1 in the flamethrower, for 300u fuel. Back tank is 500u. 200 extra fuel, (that people make up for by leaving welderpacks at FOB), but in exchange you’re losing a back slot, and a suit slot if you want to store the flamethrower. 

But the mindless pushing you’re intending to stop with this intentional nerf has made it so whenever you drop your gun, even with a mag harness, you have to wait for 3 timers to be able to fire again, the wield timer, the connection timer (that can be interrupted again), and the fire delay, which in practice isn’t much worth both giving up a slot that could be taken by a heavier, secondary weapon instead of being limited to pistols and the T-34, and a less punishing wield time to boot.


sidenote: shouldn’t pushing be encouraged, instead of more ways of punishing marines for doing so?

Another sidenote: blue flame price is as other people have stated, much too high for how actually effective it is.

**Comment by BraveMole** [comment]
> But the mindless pushing you’re intending to stop with this intentional nerf has made it so whenever you drop your gun, even with a mag harness, you have to wait for 3 timers to be able to fire again, the wield timer, the connection timer (that can be interrupted again), and the fire delay, which in practice isn’t much worth both giving up a slot that could be taken by a heavier, secondary weapon instead of being limited to pistols and the T-34, and a less punishing wield time to boot.

Allright, taking into account. The "mindless push" it allows was brought by several people, that's why I implemented that nerf.  This will stay untel i get more inputs

> Another sidenote: blue flame price is as other people have stated, much too high for how actually effective it is.
I'll propose 100 points instead of 150 points. Keep in mind that many people see blue flame as something that should not be in the game, so compromise must be made

**Comment by Knuff-les** [comment]
> > Usually, fuel loadout is 3 tanks + 1 in the flamethrower, for 300u fuel. Back tank is 500u. 200 extra fuel, (that people make up for by leaving welderpacks at FOB), but in exchange you’re losing a back slot, and a suit slot if you want to store the flamethrower.
> > But the mindless pushing you’re intending to stop with this intentional nerf has made it so whenever you drop your gun, even with a mag harness, you have to wait for 3 timers to be able to fire again, the wield timer, the connection timer (that can be interrupted again), and the fire delay, which in practice isn’t much worth both giving up a slot that could be taken by a heavier, secondary weapon instead of being limited to pistols and the T-34, and a less punishing wield time to boot.
> 
> Allright, taking into account. The "mindless push" it allows was brought by several people, that's why I implemented that nerf. You're losing a back slot but gaining a belt slot, at least that's what happens in my loadout, so you can still have a secondary without issues
> 
> > Another sidenote: blue flame price is as other people have stated, much too high for how actually effective it is.
> 
> I'll propose 100 points instead of 150 points. Keep in mind that many people see blue flame as something that should not be in the game, so compromise must be made

I understand about that, makes sense to me

and yeah, you gain a belt slot, but belt slots are limited to pistols (most are useless) and the double barrel (regarded as a joke). But it is your MR, so it’s up to you where you go with it. Best of luck

**Comment by SkeletalElite** [comment]
> > If you drop your flamer and have a mag harness it unlinks from the fuel on back
> 
> Yeah it's intended, to nerf it a bit. That prevents mindless flamer push that never stop

Don't think the back fuel tank will be worth using then unless you can attach a fuel tank to the flamethrower at the same time so that you can use the tank on the thrower itself as a backup for when you get stunned.

Just having a welder fuel kit on your back will likely be more effective. Yeah you your fuel capacity before reloading isn't the entirety of the fuel you carry on you if you do that, but the burden of reattaching your flamer fuel every time you get stunned with it out is just too oppressive for that minor advantage as it is very easy to find a moment to reload behind the wall of flames you have likely created that has facilitated your need to reload.

Just to clarify my usual loadout for flamer is flamer on suit, med belt on belt, welder kit on back (I just reload my single fuel tank from the welder kit on my back) and sidearm pouch plus pistol mag pouch for ammo + black webbing with tools

**Comment by AGasCanister** [comment]
> > If you drop your flamer and have a mag harness it unlinks from the fuel on back
> 
> Yeah it's intended, to nerf it a bit. That prevents mindless flamer push that never stop

Damn, it's gonna be so fun playing with flamer now, can't wait to be stunned by 50 different things and needing to wait an extra second to connect the flamer specially if the pyro backpack (which I imagine is the M-22 fuel kit) is gonna get removed

**Comment by SkeletalElite** [comment]
> Damn, it's gonna be so fun playing with flamer now, can't wait to be stunned by 50 different things and needing to wait an extra second to connect the flamer specially if the pyro backpack (which I imagine is the M-22 fuel kit) is gonna get removed

The fuel kit isn't removed, so you can continue to use them and not use the direct connection like I imagine most are going to do.

File: code/game/objects/items.dm [hunk 12]
```diff
@@ -244,6 +244,9 @@
 	if(flags_item & DELONDROP)
 		qdel(src)
 
+///Called whenever an item is unequipped to a new loc (IE, not when the item ends up in the hands)
+/obj/item/proc/removed_from_inventory(mob/user)
```
Comment (SplinterGP) [inline code/game/objects/items.dm:248]: that happens when dropped ? and it sends a signal ?

**Comment by BraveMole** [comment]
Capture
Happens here

File: code/game/objects/items.dm (hunk 12 above)
Comment (BraveMole) [reply code/game/objects/items.dm:248]: Capture
Happens here

**Comment by isotope436** [comment]
Love the idea and I can see the uses sadly I didn't get my hands on it when it was test merged so I cant provide any input other than I'm kind of sad to not see B fuel in the list (green fire for those of you unware of the actual names) but I will definitely see uses in this. Also iv mained flamer for ages never once have I used a side arm other than a T19 if I feel like solo pushing soooo not really and issue.

File: code/modules/projectiles/guns/flamer.dm [hunk 13]
```diff
@@ -389,19 +474,29 @@
...
 	update_icon()
 	return TRUE
 
+/obj/item/weapon/gun/flamer/marinestandard/wield(mob/user)//Auto linking
+	if (!current_mag)
+		var/mob/living/carbon/human/human_user = user
+		var/obj/item/ammo_magazine/flamer_tank/backtank/back_tank = human_user.get_type_in_slots(/obj/item/ammo_magazine/flamer_tank/backtank)
+		if (back_tank)
+			attach_fueltank(user, back_tank)
+	.=..()
```
Comment (TiviPlus) [inline code/modules/projectiles/guns/flamer.dm:499]: Return ..()

File: code/modules/projectiles/guns/flamer.dm [hunk 14]
```diff
@@ -414,12 +509,58 @@
...
+		if (!hydro_active)
+			light_pilot(user,TRUE)
+		playsound(user, reload_sound, 25, 1, 5)
+		update_icon(user)
+		var/obj/screen/ammo/A = user.hud_used.ammo
+		A.update_hud(user)
+		return
+	else
+		to_chat(user, "<span class='warning'>Your action was interrupted!</span>")
+		return
```
Comment (TiviPlus) [inline code/modules/projectiles/guns/flamer.dm:554]: Bad else

File: code/modules/projectiles/magazines/flamer.dm [hunk 15]
```diff
@@ -48,6 +48,55 @@
...
+	max_rounds = 500
+	current_rounds = 500
+	reload_delay = 1 SECONDS
+	gun_type = /obj/item/weapon/gun/flamer
+	var/obj/item/weapon/gun/flamer/attached_flamer = null
+
+
+/obj/item/ammo_magazine/flamer_tank/backtank/attackby(obj/item/I, mob/user, params)
+	. = ..()
+	if(istype(I, /obj/item/weapon/gun/flamer/))
```
Comment (TiviPlus) [inline code/modules/projectiles/magazines/flamer.dm:66]: Trailing /

File: code/modules/projectiles/magazines/flamer.dm [hunk 16]
```diff
@@ -48,6 +48,55 @@
...
+				else if (attached_flamer)
+					to_chat(user, "<span class='warning'>This fuel tank is already attached to something</span>")
+					return
+				else
+					FLT.attach_fueltank(user,src)
+					return
+			else
+				to_chat(user, "<span class='warning'>You must equip or hold this fuel tank to be able to link it to a flamer</span>")
+				return
+		return
```
Comment (TiviPlus) [inline code/modules/projectiles/magazines/flamer.dm:83]: Bad elses

**Review by TiviPlus — CHANGES_REQUESTED** [review]
Should be it I think

File: code/modules/projectiles/guns/flamer.dm [hunk 17]
```diff
@@ -414,12 +509,57 @@
...
+	return
 
 /obj/item/weapon/gun/flamer/marinestandard/Fire(atom/target, mob/living/user, params, reflex)
 	if(active_attachable && istype(active_attachable, /obj/item/attachable/hydro_cannon) && (world.time > last_use + 10))
 		extinguish(target,user) //Fire it.
+		water_count -=7//reagents is not updated in this proc, we need water_count for a updated HUD
 		last_fired = world.time
 		last_use = world.time
+		var/obj/screen/ammo/A = user.hud_used.ammo
+		A.update_hud(user)
```
Comment (TiviPlus) [inline code/modules/projectiles/guns/flamer.dm:562]:
```suggestion
		user.hud_used.ammo.update_hud(user)
```

File: code/modules/projectiles/magazines/flamer.dm [hunk 18]
```diff
@@ -48,6 +48,56 @@
...
+
+/obj/item/ammo_magazine/flamer_tank/backtank/removed_from_inventory(mob/user) //Dropping the tank should unlink it from the flamer
+	. = ..()
+	var/mob/living/carbon/human/humanuser = user
+	if (istype(humanuser))
+		if(attached_flamer)
+			attached_flamer.detach_fueltank(user,FALSE)
+			return
+		return
+	return
```
Comment (TiviPlus) [inline code/modules/projectiles/magazines/flamer.dm:94]: Early return

**Review by TiviPlus — APPROVED** [review]

File: code/modules/projectiles/magazines/flamer.dm [hunk 19]
```diff
@@ -48,6 +48,55 @@
...
+					to_chat(user, "<span class='warning'>This fuel tank is already attached to something</span>")
+					return
+				
+				FLT.attach_fueltank(user,src)
+				return
+			
+			to_chat(user, "<span class='warning'>You must equip or hold this fuel tank to be able to link it to a flamer</span>")
+			return
+		return
+
```
Comment (TiviPlus) [inline code/modules/projectiles/magazines/flamer.dm:85]: Early return

File: code/_onclick/hud/screen_objects.dm [hunk 20]
```diff
@@ -732,7 +732,8 @@
 			spawn(20)
```
Comment (SplinterGP) [inline code/_onclick/hud/screen_objects.dm:732]: spawn yikes, one day we'll remove all of this

**Review by TiviPlus — APPROVED** [review]
Good enough

File: code/modules/projectiles/guns/flamer.dm [hunk 21]
```diff
@@ -117,37 +129,102 @@
 				to_chat(user, "<span class='notice'>You begin reloading [src]. Hold still...</span>")
 				if(do_after(user,magazine.reload_delay, TRUE, src, BUSY_ICON_GENERIC))
 					replace_magazine(user, magazine)
+					light_pilot(user,TRUE)
 				else
 					to_chat(user, "<span class='warning'>Your reload was interrupted!</span>")
 					return
 			else
 				replace_magazine(user, magazine)
+				light_pilot(user,TRUE)
```
Comment (SplinterGP) [inline code/modules/projectiles/guns/flamer.dm:138]:
```suggestion
				if(!do_after(user,magazine.reload_delay, TRUE, src, BUSY_ICON_GENERIC))
					to_chat(user, "<span class='warning'>Your reload was interrupted!</span>")
					return
			replace_magazine(user, magazine)
			light_pilot(user,TRUE)
```

File: code/modules/projectiles/guns/flamer.dm [hunk 22]
```diff
@@ -117,37 +129,102 @@
...
+		replace_ammo(user, fueltank)
+		light_pilot(user,TRUE)
+		playsound(user, reload_sound, 25, 1, 5)
+		update_icon(user)
+		var/obj/screen/ammo/A = user.hud_used.ammo
+		A.update_hud(user)
+		return
+	else
+		to_chat(user, "<span class='warning'>Your action was interrupted!</span>")
+		return
```
Comment (SplinterGP) [inline code/modules/projectiles/guns/flamer.dm:200]: early return, 
if(!do_after())
   to_chat()
   return

File: code/modules/projectiles/guns/flamer.dm [hunk 23]
```diff
@@ -389,19 +474,29 @@
 				to_chat(user, "<span class='notice'>You begin reloading [src]. Hold still...</span>")
 				if(do_after(user,magazine.reload_delay, TRUE, src, BUSY_ICON_GENERIC))
 					replace_magazine(user, magazine)
+					light_pilot(user,TRUE)
 				else
 					to_chat(user, "<span class='warning'>Your reload was interrupted!</span>")
 					return
 			else
 				replace_magazine(user, magazine)
+				light_pilot(user,TRUE)
```
Comment (SplinterGP) [inline code/modules/projectiles/guns/flamer.dm:483]: same as the other reload(), use do_after as early return

File: code/game/objects/items.dm [hunk 24]
```diff
@@ -244,6 +244,10 @@
...
```
Comment (SplinterGP) [inline code/game/objects/items.dm:248]: Add a signal for this proc to send
COMSIG_REMOVED_FROM_INVENTORY or something, would be really good if anyone wanted to know when a item is removed from inventory.

**Review by SplinterGP — CHANGES_REQUESTED** [review]

File: code/modules/projectiles/magazines/flamer.dm [hunk 25]
```diff
@@ -48,6 +48,58 @@
...
+
+/obj/item/ammo_magazine/flamer_tank/backtank/attackby(obj/item/I, mob/user, params)
+	. = ..()
+	if(!istype(I, /obj/item/weapon/gun/flamer))
+		return	
+	var/obj/item/weapon/gun/flamer/FLT = I
+	var/mob/living/carbon/human/humanuser = user
+	if (!istype(humanuser))
+		return
```
Comment (SplinterGP) [inline code/modules/projectiles/magazines/flamer.dm:71]: istype before

File: code/modules/projectiles/guns/flamer.dm [hunk 26]
```diff
@@ -414,12 +499,57 @@
 	to_chat(user, "<span class='notice'>Its hydro cannon contains [reagents.get_reagent_amount(/datum/reagent/water)]/[reagents.maximum_volume] units of water!</span>")
 
 
+/obj/item/weapon/gun/flamer/marinestandard/unique_action(mob/user)
+	var/obj/item/attachable/hydro_cannon/hydro = under
+	if(istype(hydro))
```
Comment (SplinterGP) [inline code/modules/projectiles/guns/flamer.dm:504]:
```suggestion
	if(!istype(under, /obj/item/attachable/hydro_cannon)
		return
	var/obj/item/attachable/hydro_cannon/hydro = under
```
then backspace one indent from the rest

File: code/modules/projectiles/guns/flamer.dm [hunk 27]
```diff
@@ -111,43 +123,103 @@
...
+	light_pilot(user,FALSE)
+	update_icon(user)
+	var/obj/screen/ammo/A = user.hud_used.ammo
+	A.update_hud(user)
+	
+
+/obj/item/weapon/gun/flamer/removed_from_inventory(mob/user)
+	. = ..()
+	var/mob/living/carbon/human/humanuser = user
+	if (istype(current_mag,/obj/item/ammo_magazine/flamer_tank/backtank/)) //Dropping the flamer unlink it from the tank
```
Comment (SplinterGP) [inline code/modules/projectiles/guns/flamer.dm:217]:
```suggestion
	if (istype(current_mag,/obj/item/ammo_magazine/flamer_tank/backtank)) //Dropping the flamer unlink it from the tank
```

File: code/modules/projectiles/gun_attachables.dm [hunk 28]
```diff
@@ -1697,18 +1697,20 @@ inaccurate. Don't worry if force is ever negative, it won't runtime.
...
 			to_chat(user, "<span class='notice'>You are now using [src].</span>")
 		master_gun.active_attachable = src
 		icon_state += "-on"
-
+		activated = TRUE
 	for(var/X in master_gun.actions)
 		var/datum/action/A = X
 		A.update_button_icon()
-	return TRUE
+	return activated
```
Comment (SplinterGP) [inline code/modules/projectiles/gun_attachables.dm:1716]: just use . instead, byond default return value

**Review by SplinterGP — CHANGES_REQUESTED** [review]

File: code/__DEFINES/dcs/signals.dm [hunk 29]
```diff
@@ -208,7 +208,8 @@
 #define COMSIG_ITEM_EQUIPPED_TO_SLOT "item_equip_to_slot"			//from base of obj/item/equipped(): (/mob/equipper)
 #define COMSIG_ITEM_EQUIPPED_NOT_IN_SLOT "item_equip_not_in_slot"	//from base of obj/item/equipped(): (/mob/equipper, slot)
 #define COMSIG_ITEM_UNEQUIPPED "item_unequip"						//from base of obj/item/unequipped(): (/mob/unequipper, slot)
-#define COMSIG_ITEM_DROPPED "item_drop"							//from base of obj/item/dropped(): (mob/user)
+#define COMSIG_ITEM_DROPPED "item_drop"								//from base of obj/item/dropped(): (mob/user)
+#define COMSIG_ITEM_REMOVED_INVENTORY "item_removed_inventory"		//from base of obj/item/removed_from_inventory() :(mov/user)
```
Comment (TiviPlus) [inline code/__DEFINES/dcs/signals.dm:212]: mov

File: code/game/objects/items.dm [hunk 30]
```diff
@@ -244,6 +244,11 @@
...
+	SEND_SIGNAL(src, COMSIG_ITEM_REMOVED_INVENTORY, user)
+	return
```
Comment (TiviPlus) [inline code/game/objects/items.dm:250]: For the record adding new signals is fine as long as they're in generic procs but try not to overdo it

File: code/modules/projectiles/guns/flamer.dm [hunk 31]
```diff
@@ -111,43 +123,103 @@
...
+	A.update_hud(user)
+
+
+/**Proced when unlinking the back fuel tank, making the flamer unlit and unable to fire
+ * mob/user if not null, will allow to play sound and update icons / hud
+ * volontary if TRUE, will span a notice describing the action
+ */
+/obj/item/weapon/gun/flamer/proc/detach_fueltank(mob/user,volontary = TRUE)
+	var/obj/item/ammo_magazine/flamer_tank/backtank/fueltank = current_mag
+	current_mag = null
```
Comment (TiviPlus) [inline code/modules/projectiles/guns/flamer.dm:203]:
```suggestion
/obj/item/weapon/gun/flamer/proc/detach_fueltank(mob/user, volontary = TRUE)
	var/obj/item/ammo_magazine/flamer_tank/backtank/fueltank = current_mag
	current_mag = null
```

File: code/modules/projectiles/guns/flamer.dm [hunk 32]
```diff
@@ -111,43 +123,103 @@
...
+	update_icon(user)
+	var/obj/screen/ammo/A = user.hud_used.ammo
+	A.update_hud(user)
+
+
+/**Proced when unlinking the back fuel tank, making the flamer unlit and unable to fire
+ * mob/user if not null, will allow to play sound and update icons / hud
+ * volontary if TRUE, will span a notice describing the action
+ */
+/obj/item/weapon/gun/flamer/proc/detach_fueltank(mob/user,volontary = TRUE)
```
Comment (TiviPlus) [inline code/modules/projectiles/guns/flamer.dm:201]: voluntary not volontary

File: code/modules/projectiles/guns/flamer.dm [hunk 33]
```diff
@@ -111,43 +123,103 @@
...
+
+/obj/item/weapon/gun/flamer/removed_from_inventory(mob/user)
+	. = ..()
+	var/mob/living/carbon/human/humanuser = user
+	if (istype(current_mag,/obj/item/ammo_magazine/flamer_tank/backtank)) //Dropping the flamer unlink it from the tank
+		var/obj/item/ammo_magazine/flamer_tank/backtank/backfueltank = current_mag;
+		backfueltank.attached_flamer=null
+		current_mag = null 
+		light_pilot(null,FALSE) 
+	update_icon(humanuser)
```
Comment (TiviPlus) [inline code/modules/projectiles/guns/flamer.dm:222]: Why pass as an arg?

File: code/modules/projectiles/guns/flamer.dm [hunk 34]
```diff
@@ -359,12 +436,15 @@
 		/obj/item/attachable/hydro_cannon,
 	)
 	var/last_use
+	var/hydro_active
+	var/water_count
```
Comment (TiviPlus) [inline code/modules/projectiles/guns/flamer.dm:440]: Is this a bool and a integer?
If yes please define as such
(var/bool = FALSE, var/integ = 0)

File: code/modules/projectiles/magazines/flamer.dm [hunk 35]
```diff
@@ -48,6 +48,60 @@
...
+	name = "back fuel tank"
+	desc = "A specialized fuel tank for use with the TL-84 flamethrower and M240A1 incinerator unit."
+	icon_state = "flamethrower_tank"
+	flags_equip_slot = ITEM_SLOT_BACK
+	w_class = WEIGHT_CLASS_BULKY
+	max_rounds = 500
+	current_rounds = 500
+	reload_delay = 1 SECONDS
+	gun_type = /obj/item/weapon/gun/flamer
+	var/obj/item/weapon/gun/flamer/attached_flamer = null
```
Comment (TiviPlus) [inline code/modules/projectiles/magazines/flamer.dm:61]:
```suggestion
	var/obj/item/weapon/gun/flamer/attached_flamer
```

**Review by TiviPlus — CHANGES_REQUESTED** [review]

File: code/modules/projectiles/guns/flamer.dm [hunk 36]
```diff
@@ -414,12 +500,58 @@
...
+		replace_ammo(user, fueltank)
+		if (!hydro_active)
+			light_pilot(user,TRUE)
+		playsound(user, reload_sound, 25, 1, 5)
+		update_icon(user)
+		var/obj/screen/ammo/A = user.hud_used.ammo
+		A.update_hud(user)
+		return
+	to_chat(user, "<span class='warning'>Your action was interrupted!</span>")
+	return
```
Comment (TiviPlus) [inline code/modules/projectiles/guns/flamer.dm:545]:
```suggestion
	if(!do_after(user,fueltank.reload_delay, TRUE, src, BUSY_ICON_GENERIC))
		to_chat(user, "<span class='warning'>Your action was interrupted!</span>")
		return
	if (current_mag)
		if(istype(current_mag,/obj/item/ammo_magazine/flamer_tank/backtank))
			detach_fueltank(user,FALSE)
		else
			user.put_in_hands(current_mag)//We remove the fuel tank if there is one
	current_mag = fueltank
	fueltank.attached_flamer = src
	replace_ammo(user, fueltank)
	if (!hydro_active)
		light_pilot(user,TRUE)
	playsound(user, reload_sound, 25, 1, 5)
	update_icon(user)
	var/obj/screen/ammo/A = user.hud_used.ammo
	A.update_hud(user)
```

**Review by TiviPlus — APPROVED** [review]

**Review by TiviPlus — APPROVED** [review]

**Pull request merged**